cmd/promoted-image-governor/main.go

```diff
@@ -0,0 +1,314 @@
+				delete(tagsToCheck, tag)
+			}
+		}
+	}
+	mirroredTags, err := mirroredTagsByReleaseController(ctx, client, imageStreamRefs)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("failed to get the mirrored tags by release controller: %w", err))
+	}
+	for _, tag := range mirroredTags {
+		delete(tagsToCheck, tag)
+	}
+	if len(errs) > 0 {
+		return nil, utilerrors.NewAggregate(errs)
+	}
+	var tagsToDelete []api.ImageStreamTagReference
+	for tag := range tagsToCheck {
+		tagsToDelete = append(tagsToDelete, tag)
+	}
+
+	// sort for tests
+	sort.Slice(tagsToDelete, func(i, j int) bool {
+		if tagsToDelete[i].Namespace != tagsToDelete[j].Namespace {
+			return tagsToDelete[i].Namespace < tagsToDelete[j].Namespace
+		}
+		if tagsToDelete[i].Name != tagsToDelete[j].Name {
+			return tagsToDelete[i].Name < tagsToDelete[j].Name
+		}
+		return tagsToDelete[i].Tag < tagsToDelete[j].Tag
+	})
+
+	return tagsToDelete, nil
+}
+
+func mirroredTagsByReleaseController(ctx context.Context, client ctrlruntimeclient.Client, refs []ImageStreamRef) ([]api.ImageStreamTagReference, error) {
+	var ret []api.ImageStreamTagReference
+	for _, ref := range refs {
+		imageStream := &imagev1.ImageStream{}
+		if err := client.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: "ocp", Name: ref.Name}, imageStream); err != nil {
+			return nil, fmt.Errorf("could not get image stream %s in namespace ocp", ref.Name)
+		}
+		excludedTags := sets.NewString(ref.ExcludeTags...)
+		for _, tag := range imageStream.Status.Tags {
+			if !excludedTags.Has(tag.Tag) {
+				ret = append(ret, api.ImageStreamTagReference{
+					Namespace: ref.Namespace,
+					Name:      ref.Name,
+					Tag:       tag.Tag,
+				})
+			}
+		}
+	}
+
+	return ret, nil
+}
+
+type ReleaseControllerMirrorConfig struct {
+	Publish Publish `json:"publish"`
+}
+
+type Publish struct {
+	MirrorToOrigin MirrorToOrigin `json:"mirror-to-origin"`
+}
+
+type MirrorToOrigin struct {
+	ImageStreamRef ImageStreamRef `json:"imageStreamRef"`
+}
+
+type ImageStreamRef struct {
+	Namespace   string   `json:"namespace"`
+	Name        string   `json:"name"`
+	ExcludeTags []string `json:"excludeTags,omitempty"`
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	opts := parseOptions()
+
+	if err := opts.validate(); err != nil {
+		logrus.WithError(err).Fatal("failed to validate the option")
+	}
+
+	if err := addSchemes(); err != nil {
+		logrus.WithError(err).Fatal("failed to add schemes")
+	}
+
+	abs, err := filepath.Abs(opts.releaseControllerMirrorConfigDir)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to determine absolute release controller mirror config path")
+	}
+
+	var imageStreamRefs []ImageStreamRef
+	if err := filepath.Walk(abs,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				logrus.WithField("source-file", path).WithError(err).Error("Failed to walk release controller mirror config dir")
+				return err
+			}
+			if strings.HasSuffix(path, "-ci.json") {
```

Comment:
Why only CI configs? Do the others have mirroring options? Would it be wrong to read all mirroring options?

Reply:
I checked. All the mirroring options are in the `*-ci.json` files.
But yeah, changed to 

```
if strings.HasSuffix(path, ".json") {
```